### PR TITLE
Implement SpannerExchangeStepsService.streamExchangeSteps

### PR DESCRIPTION
**Pull request by yunyeng**
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)

### CODE REVIEW

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 83 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnfCsEL0QbUR2deAIvd:-MnfCsEL0QbUR2deAIve:brj464k) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/15b201ba0e87a59b8ea7cff17abe2bda05a4a732/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L83)):*
> ```Kotlin
> 
>     if (conjuncts.isEmpty()) {
>       return
> ```

It looks like this is impossible because the provider will always cause a conjunct to be added. Can you replace this with `check(conjects.isNotEmpty())`?
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/ExchangeStepsServiceTest.kt, line 447 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnfDk1G5X4UCJzRczFN:-MnfDk1G5X4UCJzRczFO:bb5jxtx) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/15b201ba0e87a59b8ea7cff17abe2bda05a4a732/src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/ExchangeStepsServiceTest.kt#L447)):*
> ```Kotlin
>   }
> 
>   private suspend fun createRecurringExchange(recExchange: RecurringExchange? = null) {
> ```

Can you default it to `RECURRING_EXCHANGE` instead of making it nullable? `recurringExchange: RecurringExchange = RECURRING_EXCHANGE`

**Review by yunyeng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: 2 of 4 files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @efoxepstein, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 83 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnfCsEL0QbUR2deAIvd:-MnfaAPO6lMaae2BZmaW:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/15b201ba0e87a59b8ea7cff17abe2bda05a4a732/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L83)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

It looks like this is impossible because the provider will always cause a conjunct to be added. Can you replace this with `check(conjects.isNotEmpty())`?
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/ExchangeStepsServiceTest.kt, line 447 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnfDk1G5X4UCJzRczFN:-Mnfa4Zj4bQoflREUCCI:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/15b201ba0e87a59b8ea7cff17abe2bda05a4a732/src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/ExchangeStepsServiceTest.kt#L447)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Can you default it to `RECURRING_EXCHANGE` instead of making it nullable? `recurringExchange: RecurringExchange = RECURRING_EXCHANGE`
</blockquote></details>

Done.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @duliomatos1 and @stevenwarejones)

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 90 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mngk58N4JFg1MJEIDBX:-Mngk58N4JFg1MJEIDBY:b8idghr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/bbd1ee6b0bd841afd4619c13cef56cc78ca0a869/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L90)):*
> ```Kotlin
>     check(conjuncts.isNotEmpty())
>     appendClause("WHERE ")
>     append(conjuncts.joinToString(" AND "))
> ```

You also need `ORDER BY UpdateTime ASC` to make sure you don't miss things.

**Review by yunyeng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1, @efoxepstein, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 90 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mngk58N4JFg1MJEIDBX:-MnglYK87B_-A2WQls5s:b-xcm6ge) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/bbd1ee6b0bd841afd4619c13cef56cc78ca0a869/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L90)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

You also need `ORDER BY UpdateTime ASC` to make sure you don't miss things.
</blockquote></details>

Its done at line 40.

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 90 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mngk58N4JFg1MJEIDBX:-MngqZ3DAarLHcyCbvNz:b-ha3uet) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/bbd1ee6b0bd841afd4619c13cef56cc78ca0a869/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L90)):*
<details><summary><i>Previously, yunyeng (Yunus Yen) wrote…</i></summary><blockquote>

Its done at line 40.
</blockquote></details>

Oh, then I think you need to either flip the order of them (order by UpdateTime first) OR the page token needs to be <Date, UpdateTime> and you need to be very careful to update the dates filter each time.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @duliomatos1 and @stevenwarejones)

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1, 1 of 2 files at r2, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1 and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 66 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnkT0KpC8ixzOhpaqrQ:-MnkT0KpC8ixzOhpaqrR:ba8o5qw) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L66)):*
> ```Kotlin
>     bind(Params.EXTERNAL_PROVIDER_ID to filter.provider.externalId)
> 
>     if (filter.externalRecurringExchangeId != 0L) {
> ```

is the default on these filters 0? I know only one recurringId would ever be equal to zero but this feels a little gross. Can we check for null here? or `filter.hasExternalRecurringExchangeId`?
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto, line 71 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnkSg1i9xe9AM33iiEe:-MnkSg1i9xe9AM33iiEf:b-qawp9l) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto#L71)):*
> ```protobuf
>     repeated google.type.Date dates = 3;
>     repeated ExchangeStep.State states = 4;
>     google.protobuf.Timestamp updated_after = 5;
> ```

do we need an `updated_before`?

**Review by efoxepstein — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 66 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnkT0KpC8ixzOhpaqrQ:-MnlLQ8_1_dZai_f7_WB:b-mb3ldp) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L66)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

is the default on these filters 0? I know only one recurringId would ever be equal to zero but this feels a little gross. Can we check for null here? or `filter.hasExternalRecurringExchangeId`?
</blockquote></details>

All proto integer fields default to zero. Proto3 doesn't have has-bits unless you specify the `optional` keyword before the field.

But since zero is [*illegal*](https://github.com/world-federation-of-advertisers/common-jvm/blob/main/src/main/kotlin/org/wfanet/measurement/common/identity/Identifiers.kt#L29), we have two choices here:

1. Assume that zero indicates no RecurringExchange (implemented)
2. Add a has-bit to the `external_recurring_exchange_id` field in the filter and implement logic like `if (filter.hasExternalRecurringExchangeId) { require(filter.externalRecurringExchangeId > 0) { "Invalid external_recurring_exchange_id: ${filter.externalRecurringExchangeId}" } ... }`

I don't care super strongly which option we go with.

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto, line 66 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mnll8Mv4AHcHZX79K1q:-Mnll8Mv4AHcHZX79K1r:b4n6lsd) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto#L66)):*
> ```protobuf
>   // Restricts results to those matching all of the given criteria. Repeated
>   // fields are treated as disjunctions.
>   message Filter {
> ```

Based on latest comments on https://github.com/world-federation-of-advertisers/cross-media-measurement-api/pull/70, we'll likely need a few more filters here:

1. `repeated Provider recurring_exchange_participants`
2. `fixed64 external_exchange_id`

**Review by stevenwarejones — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1, @efoxepstein, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 66 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnkT0KpC8ixzOhpaqrQ:-Mnly-kc6TdcYvkCR3ZD:b-s1b1j0) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L66)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

All proto integer fields default to zero. Proto3 doesn't have has-bits unless you specify the `optional` keyword before the field.

But since zero is [*illegal*](https://github.com/world-federation-of-advertisers/common-jvm/blob/main/src/main/kotlin/org/wfanet/measurement/common/identity/Identifiers.kt#L29), we have two choices here:

1. Assume that zero indicates no RecurringExchange (implemented)
2. Add a has-bit to the `external_recurring_exchange_id` field in the filter and implement logic like `if (filter.hasExternalRecurringExchangeId) { require(filter.externalRecurringExchangeId > 0) { "Invalid external_recurring_exchange_id: ${filter.externalRecurringExchangeId}" } ... }`

I don't care super strongly which option we go with.
</blockquote></details>

i don't think our version of proto we are using supports the optional keyword yet. I do like option 2 better but I'm fine with option 1 for now.

**Review by efoxepstein — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 66 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnkT0KpC8ixzOhpaqrQ:-MnlyY1x2Rx29rlrg4Qk:bo7onco) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L66)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

i don't think our version of proto we are using supports the optional keyword yet. I do like option 2 better but I'm fine with option 1 for now.
</blockquote></details>

VID code uses it extensively, e.g. https://github.com/world-federation-of-advertisers/virtual-people-common/blob/main/src/main/proto/wfa/virtual_people/common/event.proto#L75

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r6, 2 of 3 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 4 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 49 at r8](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mo0k-0NEgC80sOEgN_-:-Mo0k-0NEgC80sOEgN_0:b-i6zl7q) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/b9ad5c2562f915a6f8f8442ca1363d444b146094/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L49)):*
> ```Kotlin
> 
>     if (filter.hasStepProvider()) {
>       conjuncts.add(providerFilter(filter.stepProvider, Params.EXTERNAL_PROVIDER_ID))
> ```

I think this one is wrong now that we updated the joins. I think you want to do something like:

```
conjuncts.add("ExchangeSteps.WhateverProviderId = @whateverProviderId")
```

In other words, `stepProvider` needs to reference ExchangeSteps.WhateverProviderId, not RecurringExchanges.WhateverProviderId.

Can you add a test case for this one that fails as it is now and then fix it?

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 5 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 58 at r9](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mo0nrw2CStDOBrVoUUp:-Mo0nrw2CStDOBrVoUUq:b-o418vs) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/1914a1fb30b03043becfe67130e1a5d447637d3b/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L58)):*
> ```Kotlin
>         RecurringExchanges.ExternalRecurringExchangeId IN
>         UNNEST(@${Params.EXTERNAL_RECURRING_EXCHANGE_ID})
>         """.trimIndent()
> ```

Probably better to just use string concatenation here to avoid weird SQL formatting:

```
conjuncts.add(
  "RecurringExchanges.ExternalRecurringExchangeId IN " +
  "UNNEST(@${Params.EXTERNAL_RECURRING_EXCHANGE_ID})"
)
```

**Review by yunyeng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: 4 of 7 files reviewed, 4 unresolved discussions (waiting on @duliomatos1, @efoxepstein, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 49 at r8](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mo0k-0NEgC80sOEgN_-:-Mo17CryCMnCCbtiIRgZ:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/b9ad5c2562f915a6f8f8442ca1363d444b146094/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L49)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

I think this one is wrong now that we updated the joins. I think you want to do something like:

```
conjuncts.add("ExchangeSteps.WhateverProviderId = @whateverProviderId")
```

In other words, `stepProvider` needs to reference ExchangeSteps.WhateverProviderId, not RecurringExchanges.WhateverProviderId.

Can you add a test case for this one that fails as it is now and then fix it?
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt, line 58 at r9](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mo0nrw2CStDOBrVoUUp:-Mo17Ego0A2YGbsEL-ZP:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/1914a1fb30b03043becfe67130e1a5d447637d3b/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/queries/StreamExchangeSteps.kt#L58)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Probably better to just use string concatenation here to avoid weird SQL formatting:

```
conjuncts.add(
  "RecurringExchanges.ExternalRecurringExchangeId IN " +
  "UNNEST(@${Params.EXTERNAL_RECURRING_EXCHANGE_ID})"
)
```
</blockquote></details>

Done.
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto, line 66 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mnll8Mv4AHcHZX79K1q:-Mo17HWQ3A79300GeW82:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto#L66)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Based on latest comments on https://github.com/world-federation-of-advertisers/cross-media-measurement-api/pull/70, we'll likely need a few more filters here:

1. `repeated Provider recurring_exchange_participants`
2. `fixed64 external_exchange_id`
</blockquote></details>

Done.

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @stevenwarejones, and @yunyeng)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/common/ProviderFilter.kt, line 37 at r10](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-Mo19RK-0qGlGOPvFhtv:-Mo19RK-0qGlGOPvFhtw:bsices) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7007d989fffc05cc5bfbda68e7e446f2ba045b28/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/common/ProviderFilter.kt#L37)):*
> ```Kotlin
>   return when (stepProvider.type) {
>     Provider.Type.DATA_PROVIDER ->
>       "ExchangeSteps.DataProviderId = (SELECT DataProviderId FROM DataProviders " +
> ```

I don't think you need anything this complicated -- these subqueries are not too nice. Just do: `(ExchangeSteps.DataProviderId = RecurringExchanges.DataProviderId AND ${providerFilter(stepProvider, param)})`.

Alternatively, change this function to just:

```
fun stepIsOwnedByProviderTypeFilter(type: Provider.Type): String {
  return when (type) {
    Provider.Type.DATA_PROVIDER -> "ExchangeSteps.DataProviderId = RecurringExchanges.DataProviderId"
  ...
}
```

and then also add `stepProvider` to the list of `recurringExchangeParticipants`
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto, line 71 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnkSg1i9xe9AM33iiEe:-Mo1Aoe05-NDMtz6wKSm:bw23ndr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto#L71)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

do we need an `updated_before`?
</blockquote></details>

We can always add it in later. There's no immediate need.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @stevenwarejones)

**Review by yunyeng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @stevenwarejones)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto, line 71 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347#-MnkSg1i9xe9AM33iiEe:-Mo5j7EoActua_7Oo7MM:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/3e525633e1c6c65ce3dff948025fa3d92d674efb/src/main/proto/wfa/measurement/internal/kingdom/exchange_steps_service.proto#L71)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

We can always add it in later. There's no immediate need.
</blockquote></details>

Done.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 8 of 8 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @stevenwarejones)

**Review by efoxepstein — APPROVED** [review]
> Reviewed all commit messages.
> Dismissed @stevenwarejones from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/347)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @duliomatos1 and @stevenwarejones)
